File: heap(priority queue)/smallest_number_in_infinite_set.java
import java.util.PriorityQueue;;
public class smallest_number_in_infinite_set {
    private Integer minNum;
    private PriorityQueue<Integer> minHeap;
    public smallest_number_in_infinite_set() {
        minHeap = new PriorityQueue<>();
        minNum = 1;
    }
    
    public int popSmallest() {
        if(!minHeap.isEmpty()){
            return this.minHeap.poll();
        }
        int res = minNum;
        minNum++;
        return res;
    }
    
    public void addBack(int num) {
        if(num < minNum && !(minHeap.contains(num))){
            minHeap.offer(num);
        }
    }
}

File: heap(priority queue)/smallest_number_in_infinite_set.py
import heapq
class SmallestInfiniteSet:

    def __init__(self):
        self.minHeap = []
        self.next_num = 1

    def popSmallest(self) -> int:
        if self.minHeap:
            return heapq.heappop(self.minHeap)
        res = self.next_num
        self.next_num += 1
        return res

    def addBack(self, num: int) -> None:
        if num < self.next_num and num not in self.minHeap:
            heapq.heappush(self.minHeap, num)


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)